daal4py/sklearn/linear_model/_linear_0_24.py

```diff
@@ -0,0 +1,252 @@
+
+    if sp.issparse(X):
+        X_offset_scale = X_offset / X_scale
+
+        def matvec(b):
+            return X.dot(b) - b.dot(X_offset_scale)
+
+        def rmatvec(b):
+            return X.T.dot(b) - X_offset_scale * np.sum(b)
+
+        X_centered = sp.linalg.LinearOperator(shape=X.shape,
+                                              matvec=matvec,
+                                              rmatvec=rmatvec)
+
+        if y.ndim < 2:
+            out = sparse_lsqr(X_centered, y)
+            self.coef_ = out[0]
+            self._residues = out[3]
+        else:
+            # sparse_lstsq cannot handle y with shape (M, K)
+            outs = Parallel(n_jobs=n_jobs_)(
+                delayed(sparse_lsqr)(X_centered, y[:, j].ravel())
+                for j in range(y.shape[1]))
+            self.coef_ = np.vstack([out[0] for out in outs])
+            self._residues = np.vstack([out[3] for out in outs])
+    else:
+        self.coef_, self._residues, self.rank_, self.singular_ = \
+            linalg.lstsq(X, y)
+        self.coef_ = self.coef_.T
+
+    if y.ndim == 1:
+        self.coef_ = np.ravel(self.coef_)
+    self._set_intercept(X_offset, y_offset, X_scale)
+    return self
+
+def predict(self, X):
+    """Predict using the linear model
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+        Samples.
+
+    Returns
+    -------
+    C : array, shape = (n_samples,)
+        Returns predicted values.
+    """
+    is_df = is_DataFrame(X)
+    X = np.asarray(X) if not sp.issparse(X) and not is_df else X
+    good_shape_for_daal = True if X.ndim <= 1 else True if X.shape[0] > X.shape[1] else False
+    dtype = get_dtype(X)
+
+    if (sp.issparse(X) or
+            not hasattr(self, 'daal_model_') or
+            not self.fit_shape_good_for_daal_ or
+            not good_shape_for_daal or
+            not (dtype == np.float64 or dtype == np.float32) or
+            (hasattr(self, 'sample_weight_') and self.sample_weight_ is not None)):
+        logging.info("sklearn.linar_model.LinearRegression.predict: " + method_uses_sklearn)
+        return self._decision_function(X)
+    else:
+        logging.info("sklearn.linar_model.LinearRegression.predict: " + method_uses_daal)
+        X = _daal_check_array(X)
+        return _daal4py_predict(self, X)
+
+
+_fit_copy = fit
+_predict_copy = predict
+
+class LinearRegression(LinearRegression_original):
+    __doc__ = LinearRegression_original.__doc__
+
+    def __init__(self, fit_intercept=True, normalize=False, copy_X=True,
+                 n_jobs=None, positive=False):
+        super(LinearRegression, self).__init__(
+            fit_intercept=fit_intercept, normalize=normalize,
+            copy_X=copy_X, n_jobs=n_jobs, positive=positive)
+
+    def fit(self, X, y, sample_weight=None):
+        return _fit_copy(self, X, y, sample_weight=sample_weight)
+
+    def predict(self, X):
+        return _predict_copy(self, X)
+
```

Comment:
@PetrovKP, We can check sklearn version in lin_reg_0-23 wrapper in __init__, for transmission positive parameter. After that check parameter in fit method:

`class LinearRegression(LinearRegression_original):
    __doc__ = LinearRegression_original.__doc__

    def __init__(self, fit_intercept=True, normalize=False, copy_X=True,
                 n_jobs=None, positive=False):
        if (LooseVersion(sklearn_version) >= LooseVersion("0.24")):
            super(LinearRegression, self).__init__(
                fit_intercept=fit_intercept, normalize=normalize,
                copy_X=copy_X, n_jobs=n_jobs, positive=positive)
        else:
            super(LinearRegression, self).__init__(
                fit_intercept=fit_intercept, normalize=normalize,
                copy_X=copy_X, n_jobs=n_jobs)


    def fit(self, X, y, sample_weight=None):
        if self.positive == True:
            logging.info("sklearn.linar_model.LinearRegression.fit: " + method_uses_sklearn)
            return super(LinearRegression, self).fit(X, y=y, sample_weight=sample_weight)
        return _fit_copy(self, X, y, sample_weight=sample_weight)

    def predict(self, X):
        return _predict_copy(self, X)`

Reply:
I like this option more, since we bring less copy-paste.